lib/SILGen/SILGenStmt.cpp

```diff
@@ -326,6 +326,11 @@ void StmtEmitter::visitBraceStmt(BraceStmt *S) {
           // Ignore all other implicit expressions.
           continue;
         }
+      } else if (auto D = ESD.dyn_cast<Decl*>()) {
+        // Local type declarations are not unreachable because they can appear
+        // after the declared type has already been used.
+        if (isa<TypeDecl>(D))
```

Comment:
One more thing I realized: local functions (FuncDecl) need the same treatment, because they can also be forward-referenced. Do you mind adding that in this PR or a follow-up?

Reply:
Sure thing. On a related note, while testing I found out forward references in this scenario seem to [crash the compiler](https://bugs.swift.org/browse/SR-13793).